File: src/main/java/io/bloco/faker/components/Commerce.java
package io.bloco.faker.components;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;

import io.bloco.faker.FakerComponent;
import io.bloco.faker.FakerData;

public class Commerce extends FakerComponent {

    public Commerce(FakerData data) {
        super(data);
    }

    public String color() {
        return call("color.colorName");
    }

    public String department() {
        return department(3);
    }

    public String department(int max) {
        return department(max, false);
    }

    public String department(int max, boolean fixedAmount) {
        int num;

        if (fixedAmount) {
            num = max;
        } else {
            num = 1 + randomHelper.number(max);
        }

        List<String> categories = getCategories(num);

        if (num > 1) {
            return mergeCategories(categories);
        } else {
            return categories.get(0);
        }
    }

    public String productName() {
        return fetch("commerce.product_name.adjective")
                + " " + fetch("commerce.product_name.material")
                + " " + fetch("commerce.product_name.product");
    }

    public String material() {
        return fetch("commerce.product_name.material");
    }

    public BigDecimal price() {
        return price(0, 100);
    }

    public BigDecimal price(int min, int max) {
        return new BigDecimal(randomHelper.range(min, max))
                .round(new MathContext(2, RoundingMode.HALF_UP));
    }

    public String promotionCode() {
        return promotionCode(6);
    }

    public String promotionCode(int digits) {
        return fetch("commerce.promotion_code.adjective")
                + fetch("commerce.promotion_code.noun")
                + getComponent(Number.class).number(digits);
    }

    // Helpers

    private List<String> getCategories(int num) {
        List<String> categories = new ArrayList<>(num);

        while (categories.size() != num) {
            String category = fetch("commerce.department");
            if (!categories.contains(category)) {
                categories.add(category);
            }
        }

        return categories;
    }

    private String mergeCategories(List<String> categories) {
        List<String> commaCategories = categories.subList(0, categories.size() - 1);
        String commaSeparated = stringHelper.join(commaCategories, ", ");
        return commaSeparated + getSeparator() + categories.get(categories.size() - 1);
    }
}

File: lib/src/main/kotlin/io/bloco/faker/components/Commerce.kt
package io.bloco.faker.components

import io.bloco.faker.FakerComponent
import io.bloco.faker.FakerData
import java.math.BigDecimal
import java.math.RoundingMode

class Commerce(data: FakerData) : FakerComponent(data) {
    fun color(): String {
        return call("color.colorName")
    }

    @JvmOverloads
    fun department(max: Int = 3, fixedAmount: Boolean = false): String {
        val num = if (fixedAmount) max else randomHelper.number(max) + 1
        val categories = getCategories(num)
        return if (num > 1) mergeCategories(categories) else categories.first()
    }

    fun productName(): String {
        return listOf(
            fetch("commerce.product_name.adjective"),
            fetch("commerce.product_name.material"),
            fetch("commerce.product_name.product")
        ).joinToString(" ")
    }

    fun material(): String {
        return fetch("commerce.product_name.material")
    }

    @JvmOverloads
    fun price(min: Int = 0, max: Int = 100): BigDecimal {
        return BigDecimal(randomHelper.range(min, max)).setScale(2, RoundingMode.HALF_UP)
    }

    @JvmOverloads
    fun promotionCode(digits: Int = 6): String {
        return listOf(
            fetch("commerce.promotion_code.adjective"),
            fetch("commerce.promotion_code.noun"),
            getComponent(Number::class).number(digits)
        ).joinToString("")
    }

    // Helpers
    private fun getCategories(num: Int): List<String> {
        return generateSequence { fetch("commerce.department") }
            .distinct()
            .take(num)
            .toList()
    }

    private fun mergeCategories(categories: List<String>): String {
        val commaSeparated = categories.take(categories.size - 1).joinToString(", ")
        return commaSeparated + separator + categories.last()
    }
}
